### PR TITLE
First attempt to add tracers

### PR DESCRIPTION
**Pull request by francispoulin**
I have made a first attempt to add tracers into `ShallowWaterModel`, but unfortunately, have a problem with the new function I have created, `div_ucvc` to compute the divergence of the tracer, `c`, times the velocity `(u,v)`.    

When I try running a simulation I get the following output, which is a TaskFailedexcetion, and seems to be having problems with convert.

```
julia> run!(simulation)
ERROR: TaskFailedException:
MethodError: Cannot `convert` an object of type Oceananigans.AbstractOperations.BinaryOperation{Center,Center,Center,typeof(+),Oceananigans.AbstractOperations.BinaryOperation{Center,Center,Center,typeof(+),Float64,Oceananigans.AbstractOperations.BinaryOperation{Center,Center,Center,typeof(*),Field{Center,Center,Center,OffsetArrays.OffsetArray{Float64,3,Array{Float64,3}},RegularRectilinearGrid{Float64,Periodic,Periodic,Flat,OffsetArrays.OffsetArray{Float64,1,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}},NamedTuple{(:x, :y, :z),Tuple{CoordinateBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic,Nothing},BoundaryCondition{Oceananigans.BoundaryConditions.Periodic,Nothing}},CoordinateBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic,Nothing},BoundaryCondition{Oceananigans.BoundaryConditions.Periodic,Nothing}},CoordinateBoundaryConditions{Nothing,Nothing}}}},Float64,typeof(identity),typeof(identity),typeof(identity),RegularRectilinearGrid{Float64,Periodic,Periodic,Flat,OffsetArrays.OffsetArray{Float64,1,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}}},typeof(identity),typeof(identity),typeof(identity),RegularRectilinearGrid{Float64,Periodic,Periodic,Flat,OffsetArrays.OffsetArray{Float64,1,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}}},Oceananigans.AbstractOperations.BinaryOperation{Center,Center,Center,typeof(*),Field{Center,Center,Center,OffsetArrays.OffsetArray{Float64,3,Array{Float64,3}},RegularRectilinearGrid{Float64,Periodic,Periodic,Flat,OffsetArrays.OffsetArray{Float64,1,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}},NamedTuple{(:x, :y, :z),Tuple{CoordinateBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic,Nothing},BoundaryCondition{Oceananigans.BoundaryConditions.Periodic,Nothing}},CoordinateBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic,Nothing},BoundaryCondition{Oceananigans.BoundaryConditions.Periodic,Nothing}},CoordinateBoundaryConditions{Nothing,Nothing}}}},Float64,typeof(identity),typeof(identity),typeof(identity),RegularRectilinearGrid{Float64,Periodic,Periodic,Flat,OffsetArrays.OffsetArray{Float64,1,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}}},typeof(identity),typeof(identity),typeof(identity),RegularRectilinearGrid{Float64,Periodic,Periodic,Flat,OffsetArrays.OffsetArray{Float64,1,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}}} to an object of type Float64
Closest candidates are:
  convert(::Type{T}, ::Static.StaticInt{N}) where {T<:Number, N} at /home/fpoulin/.julia/packages/Static/BGJW6/src/static_implementation.jl:22
  convert(::Type{T}, ::Static.StaticFloat64{N}) where {N, T<:AbstractFloat} at /home/fpoulin/.julia/packages/Static/BGJW6/src/float.jl:26
  convert(::Type{T}, ::T) where T<:Number at number.jl:6

```

### CODE REVIEW

File: src/Models/ShallowWaterModels/shallow_water_advection_operators.jl [hunk 1]
```diff
@@ -50,3 +50,10 @@ which will end up at the location `ccc`.
     1/Vᵃᵃᶜ(i, j, k, grid) * (δxᶜᵃᵃ(i, j, k, grid, Ax_ψᵃᵃᶜ, solution.uh) + 
                              δyᵃᶜᵃ(i, j, k, grid, Ay_ψᵃᵃᶜ, solution.vh))
 end
+
+@inline function div_ucvc(i, j, k, grid, u, v, c)
```
Comment (ali-ramadhan) [inline src/Models/ShallowWaterModels/shallow_water_advection_operators.jl:54]: Would `div_Uc` be more consistent with `div_hUu` and `div_hUv` from earlier in this file?

File: src/Models/ShallowWaterModels/solution_and_tracer_tendencies.jl [hunk 2]
```diff
@@ -78,5 +78,12 @@ end
...
 
-    return ( 0.0 )
+    u = solution.uh / ℑxyᶠᶠᵃ(i, j, k, grid, solution.h)   # are these computed correctly?
+    v = solution.vh / ℑxyᶠᶠᵃ(i, j, k, grid, solution.h)
+
+    return ( 0.0
+             - div_ucvc(i, j, k, grid, u, v, c)           # are these evaluated at the correct locations?
+             + c * ∂xᶜᵃᵃ(i, j, k, grid, u)
+             + c * ∂yᵃᶜᵃ(i, j, k, grid, v) 
+            )
```
Comment (ali-ramadhan) [inline src/Models/ShallowWaterModels/solution_and_tracer_tendencies.jl:88]: I think you're mixing fields and `field[i, j, k]`?

I'm guessing you might want something similar to https://github.com/CliMA/Oceananigans.jl/blob/a211d9d0f0ddf3f66ca9fe4319f47aac1e9e6164/src/Advection/tracer_advection_operators.jl but for shallow water (might have to define `advective_tracer_flux_x` and `advective_tracer_flux_y` for shallow water)? Might end up similar to what you did for shallow water momentum advection operators above.

File: src/Models/ShallowWaterModels/shallow_water_advection_operators.jl (hunk 1 above)
Comment (francispoulin) [reply src/Models/ShallowWaterModels/shallow_water_advection_operators.jl:54]: I agree, that is a better choice.  I suppose I was looking at the mass equation where we use `div_huhv`, which is not consistent.  I wonder if thta should be changed to `div_hU`?

File: src/Models/ShallowWaterModels/solution_and_tracer_tendencies.jl (hunk 2 above)
Comment (francispoulin) [reply src/Models/ShallowWaterModels/solution_and_tracer_tendencies.jl:88]: Yes, I am certainly mixed up.  Thanks for sharing the link.  I tried to copy that before but will try it again and see if it works any better the next time around.

**Review by ali-ramadhan — APPROVED** [review]

**Comment by francispoulin** [comment]
I have included `closure` in `ShallowWaterModel` and I believe that is working correctly.  However, when I try running `examples/test_advection_shallow_water.jl` I find that I get `NaN` after the first time step.  

I know that the operator `∇_κ_∇c` does work with `Flat` as I have changed the one-dimensinal-diffusion-example and that does not have a problem.  I guess I need to figure out what is not compatable with this operator and shallow water.  If anyone has any idea what's wrong please let me know.

All tests pass and I could merge but I think I will wait to get this sorted out before I merge.  Might be good to add in a test afterwards to make sure that `ShallowWaterModel` can diffuse tracers without a problem.

**Comment by francispoulin** [comment]
I have figured out which `∇_κ_∇c` is being used and it's in `diffusion_operators.jl`.  The problem is that `V^ccc` is not prepared to deal with `Flat` and was being a zero volume, hence the `NaN`.  I had put in a change to get finite numbers but this is not a fix.  This is something we will need to fix. 

The current test example does not diffuse at all, so it's a work in progress but at least I know where the `NaNs` are coming from.

**Comment by francispoulin** [comment]
I believe I have fixed `spacings_and_areas_and_volumes.jl` to allow this example to work.  `Flat` should also work for `RegularRectilinearGrid` and `VerticallyStretchedRectilinearGrid`.  I haven't touched curvilinear grids but we should be able
to fix those in a similar way. Or we can do something else.  I guess we will see how the tests do and go from there.

**Review by ali-ramadhan — APPROVED** [review]
Looks great!

**Pull request merged**